### PR TITLE
Add differential privacy distinguisher

### PR DESCRIPTION
**Pull request by wulu473**

### CODE REVIEW

File: experiments/configs/dpd_image_classifier.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  num_train_epochs: 10.0
+  learning_rate: 0.5
+  max_physical_batch_size: 512
+  per_sample_max_grad_norm: 2.6
+  delta: 1e-5
+  lr_scheduler_gamma: 0.96
+
+game_config:
+  seed: 1920
+  num_games: 2
```
Comment (donebydan) [inline experiments/configs/dpd_image_classifier.yaml:17]: Should we remove this file / add it to gitignore?

File: experiments/scripts/debug-component.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
```
Comment (donebydan) [inline experiments/scripts/debug-component.py:1]: NIT / Reminder: this would be very useful standalone outside of the repo. No need for you to act on it here though.

File: privacy_estimates/experiments/attacks/dpd/data.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+from .attack import CanaryTrackingOptimizer
+
+
+@dataclass
+class DPDistinguishingData:
+    scores: List[float]
+    sensitivity: float
+
+    @classmethod
+    def from_opacus(cls, optimizer: DPOptimizer) -> "DPDistinguishingData":
```
Comment (donebydan) [inline privacy_estimates/experiments/attacks/dpd/data.py:16]: When you add type annotations and also check the type, maybe it is time to consider a statically typed language 😄

File: privacy_estimates/experiments/components/postprocess_dpd_data/postprocess_dpd_data.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    Instead we add the C^2 term in this script.
+    """
+    with Path(dpd_data).open() as f:
+        raw_data = load(f)
+    sensitivity = raw_data["sensitivity"]
+
+    with Path(dp_parameters).open() as f:
+        dp_parameters = load(f)
+
+    # Differential Privacy Distinguisher
```
Comment (donebydan) [inline privacy_estimates/experiments/components/postprocess_dpd_data/postprocess_dpd_data.py:33]: Do these need to be manually set with varying adversary models?

File: tests/test_experiments/test_attacks/test_dpd.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+        canary = CanaryGradient.from_optimizer(optimizer, "dirac", 1.0)
+
+        # no error
+        CanaryTrackingOptimizer(optimizer, canary)
+
+        dp_optimizer = DPOptimizer(optimizer, noise_multiplier=0.1, max_grad_norm=1, expected_batch_size=16)
+
+        with pytest.raises(TypeError):
+            CanaryTrackingOptimizer(dp_optimizer, canary)
```
Comment (donebydan) [inline tests/test_experiments/test_attacks/test_dpd.py:45]: Should we also assert the error message for ease of understanding correct behaviour here?

File: tests/test_experiments/test_attacks/test_dpd.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+        grad = next(iter(model.parameters())).grad.detach()
+
+        score = grad.flatten().dot(canary.gradient[0][0].flatten())
+
+        assert canary_tracking_optimizer.observations[0] == pytest.approx(score.item())
+
+    def get_observation(self, norm: float, method: str, mean: float, std: float, batch_size: int):
+        model = nn.Linear(10, 1, bias=False)
+        model = GradSampleModule(model)
```
Comment (donebydan) [inline tests/test_experiments/test_attacks/test_dpd.py:76]: NIT: for consistency with previous, we wrapped model in same line `GradSampleModule(nn.Linear(...))`

File: tests/test_experiments/test_attacks/test_dpd.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        x = torch.randn((batch_size, 10)) * std + mean
+        y = model(x)
+        y.mean().backward()
+
+        canary = CanaryGradient.from_optimizer(optimizer, method=method, norm=norm)
+
+        canary_tracking_optimizer = CanaryTrackingOptimizer(optimizer, canary)
+
+        # Turn off noise and clipping
+        dp_optimizer = DPOptimizer(canary_tracking_optimizer, noise_multiplier=0.0, max_grad_norm=1e4,
```
Comment (donebydan) [inline tests/test_experiments/test_attacks/test_dpd.py:91]: Why `max_grad_norm=1e4`? This will never be reached?

**Review by donebydan — CHANGES_REQUESTED** [review]

File: privacy_estimates/experiments/aml.py [hunk 8]
```diff
@@ -0,0 +1,586 @@
+import hydra
```
Comment (donebydan) [inline privacy_estimates/experiments/aml.py:1]: This file seems like a lot of boiler plate code for AML. Are these methods custom? There are no ways to pull from an AML utils library?

**Review by donebydan — CHANGES_REQUESTED** [review]

File: experiments/components/predict-with-cnn-classifier/predict_with_cnn_classifier.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+def main(args: Arguments):
+    data = load_from_disk(args.dataset)
+
+    print(f"Loaded dataset: {data.features}")
+
+    if len(data) == 0:
+        dataset = TensorDataset(torch.tensor([]), torch.tensor([]))
+    else:
+        dataset = TensorDataset(torch.tensor(data["image"]), torch.tensor(data["label"]))
+
```
[the remark this answers is not in the document]
Comment (wulu473) [reply experiments/components/predict-with-cnn-classifier/predict_with_cnn_classifier.py]: I agree, it's not great. Should be fixed

File: privacy_estimates/experiments/attacks/dpd/attack.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+            method: The method used to generate the canary gradient. Currently only "dirac" is supported.
+            norm: The norm of the canary gradient.
+            is_static: If True, the canary gradient is generated once and then cached. If False, the canary gradient is
+                          generated each time it is accessed.
+        """
+        self.shapes = shapes
+        self.method = CanaryGradientMethod(method)
+        self.is_static = is_static
+        self._cached_static_gradient = None
+        self.device = torch.device("cpu")
```
[the remark this answers is not in the document]
Comment (wulu473) [reply privacy_estimates/experiments/attacks/dpd/attack.py:35]: This is the default device. The gradient can be moved with the `to` function later.

File: privacy_estimates/experiments/attacks/dpd/attack.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+            method: The method used to generate the canary gradient. Currently only "dirac" is supported.
+            norm: The norm of the canary gradient.
+            is_static: If True, the canary gradient is generated once and then cached. If False, the canary gradient is
+                          generated each time it is accessed.
+        """
+        devices = set(p.device for param_group in optimizer.param_groups for p in param_group["params"])
+        if len(devices) != 1:
+            raise ValueError("All parameters must be on the same device")
+        shapes = [[p.shape for p in param_group["params"]] for param_group in optimizer.param_groups]
+        return CanaryGradient(shapes=shapes, method=method, norm=norm, is_static=is_static).to(devices.pop())
```
[the remark this answers is not in the document]
Comment (wulu473) [reply privacy_estimates/experiments/attacks/dpd/attack.py:52]: Updated the function to make it slightly clearer. We don't handle `self.device` in this function any more. We identify the device from the optimizer, initialize a compatible gradient on the cpu and then send it to the same device as the model.

File: privacy_estimates/experiments/attacks/dpd/attack.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+            return self._generate_dirac_canary_gradient()
+        elif self.method == CanaryGradientMethod.RANDOM:
+            return self._generate_random_canary_gradient()
+        else:
+            raise ValueError(f"Unsupported canary gradient method: {self.method}")
+
+    def _generate_dirac_canary_gradient(self) -> List[List[torch.Tensor]]:
+        """Generate a new canary gradient using the Dirac method"""
+        num_params = len(self)
+        i_rand_element = torch.randint(num_params, size=(1,))
```
[the remark this answers is not in the document]
Comment (wulu473) [reply privacy_estimates/experiments/attacks/dpd/attack.py]: Good point. Added.

File: privacy_estimates/experiments/aml.py (hunk 8 above)
Comment (wulu473) [reply privacy_estimates/experiments/aml.py:1]: I agree. This functionality isn't really provided by AML. We can think about pulling it out and having a standalone version of it at some point.

File: tests/test_experiments/test_attacks/test_dpd.py (hunk 7 above)
Comment (wulu473) [reply tests/test_experiments/test_attacks/test_dpd.py:90]: This is pretty safe. For this toy model and input gradients will be much smaller.

File: experiments/configs/dpd_image_classifier.yaml (hunk 1 above)
Comment (wulu473) [reply experiments/configs/dpd_image_classifier.yaml:17]: I think it can be useful to have a working set of hyper-parameters. It's in the experiment subdir so I hope it's clear that it's meant to be an example. I have removed the details of the workspace as they won't be helpful for anyone without access.

File: privacy_estimates/experiments/components/postprocess_dpd_data/postprocess_dpd_data.py (hunk 4 above)
Comment (wulu473) [reply privacy_estimates/experiments/components/postprocess_dpd_data/postprocess_dpd_data.py:33]: This component is called by the DPD game. We're doing this since the canary gradient attack from https://arxiv.org/pdf/2302.07956.pdf doesn't test the subsampling since the canary is inserted independently from the batch creation. Since it's independent we can do it in post. We can set the sampling probability to 1 to get the most samples from the training run but we could also set it to another value later if we wanted to.

File: experiments/README.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+We follow the attack by Nasr et al. (2023) to match the differential privacy threat model.
+Currently this does not take privacy amplification via subsampling into account and only audits a single step of DP-SGD.
+
+### Collecting membership inference data
+
+The threat model of Differentially Private Stochastic Gradient Descent (DP-SGD) assumes that the adversary has access to the training data and each gradient in the training process.
+In order to instantiate a matching adversary, we need to collect membership information during the training process.
+We provide a wrapper (`privacy_estimates.experiments.attack.dpd.CanaryTrackingOptimizer`) for a PyTorch optimizer that can be used with Opacus.
+
+`privacy_estimates.experiments.games.differential_privacy_distinguisher` contains code to run the differentially private distinguisher game.``
```
[the remark this answers is not in the document]
Comment (wulu473) [reply experiments/README.md]: Updated the readme

File: tests/test_experiments/test_attacks/test_dpd.py (hunk 5 above)
Comment (wulu473) [reply tests/test_experiments/test_attacks/test_dpd.py]: added a check for the message as well
